source/cgal/intersections.cc

```diff
@@ -685,69 +605,82 @@ namespace CGALWrappers
+    //             CGALWrappers::cgal_point_to_dealii_point<3>(tet.vertex(2)),
+    //             CGALWrappers::cgal_point_to_dealii_point<3>(tet.vertex(3))}});
+    //       }
+    //     return vertices;
+    //   }
+    // else
+    //   {
+    //     return vertices;
+    //   }
+  }
+
+  template <int dim0, int dim1, int spacedim>
+  std::vector<std::array<Point<spacedim>, dim1 + 1>>
+  compute_intersection_of_cells(
+    const typename Triangulation<dim0, spacedim>::cell_iterator &cell0,
+    const typename Triangulation<dim1, spacedim>::cell_iterator &cell1,
+    const Mapping<dim0, spacedim> &                              mapping0,
+    const Mapping<dim1, spacedim> &                              mapping1,
+    const double                                                 tol)
   {
-    Assert(ReferenceCell::n_vertices_to_type(
-             3, mapping0.get_vertices(cell0).size()) ==
-             ReferenceCells::Hexahedron,
+    // Elements have to be rectangular
+    Assert(mapping0.get_vertices(cell0).size() == std::pow(2, dim0),
            ExcNotImplemented());
-    Assert(ReferenceCell::n_vertices_to_type(
-             3, mapping1.get_vertices(cell1).size()) ==
-             ReferenceCells::Hexahedron,
+    Assert(mapping0.get_vertices(cell1).size() == std::pow(2, dim1),
            ExcNotImplemented());
-    Surface_mesh surf0, surf1, sm;
-    CGALWrappers::dealii_cell_to_cgal_surface_mesh(cell0, mapping0, surf0);
-    CGALWrappers::dealii_cell_to_cgal_surface_mesh(cell1, mapping1, surf1);
-    CGAL::Polygon_mesh_processing::triangulate_faces(surf0);
-    CGAL::Polygon_mesh_processing::triangulate_faces(surf1);
-    CGALWrappers::compute_boolean_operation(
-      surf0, surf1, CGALWrappers::BooleanOperation::compute_intersection, sm);
-    std::vector<std::array<Point<3>, 4>> vertices;
-    if (CGAL::Polygon_mesh_processing::volume(sm) > tol)
-      {
-        // Collect tetrahedrons
-        Triangulation3_inexact tria;
-        tria.insert(sm.points().begin(), sm.points().end());
-        for (const auto &c : tria.finite_cell_handles())
-          {
-            const auto &tet = tria.tetrahedron(c);
-            vertices.push_back(
-              {{CGALWrappers::cgal_point_to_dealii_point<3>(tet.vertex(0)),
-                CGALWrappers::cgal_point_to_dealii_point<3>(tet.vertex(1)),
-                CGALWrappers::cgal_point_to_dealii_point<3>(tet.vertex(2)),
-                CGALWrappers::cgal_point_to_dealii_point<3>(tet.vertex(3))}});
-          }
-        return vertices;
-      }
-    else
-      {
-        return vertices;
-      }
+
+    std::array<Point<spacedim>, int(std::pow(2, dim0))> vertices0;
+    std::array<Point<spacedim>, int(std::pow(2, dim1))> vertices1;
+    std::copy_n(mapping0.get_vertices(cell0).begin(), 8, vertices0.begin());
+    std::copy_n(mapping1.get_vertices(cell1).begin(), 4, vertices1.begin());
+
+    transform_to_cgal_vertex_order(vertices0);
+    transform_to_cgal_vertex_order(vertices1);
```

Comment:
The background is that I don't want to think about correct vertex ordering when implementing other functionality. That's why transform_to_cgal_vertex_order does nothing for 1D and 3D elements. I agree this is not perfect. What do you think about a function ```get_vertices_in_cgal_order(cell_iterator)``` instead? I would like it even better.

Reply:
It would look like this: [e487560](https://github.com/luca-heltai/dealii/pull/113/commits/e4875602146d2f77ae5008102425b332286f48e1)
(Also restricted to rectangular elements atm)

Reply:
The idea LGTM, and I agree utilities.h is the right place for it.